File: TestDPLS-iOS/TestDPLS/BLE/DplsCrypto.swift
import Foundation
import CommonCrypto
import CryptoKit

enum DplsCrypto {
    static let pbkdf2Iterations: UInt32 = 10_000

    static func deriveVerifier(password: String, salt: Data) -> Data {
        let passwordData = Data(password.utf8)
        var derived = Data(count: 32)
        let status = derived.withUnsafeMutableBytes { derivedPtr in
            passwordData.withUnsafeBytes { passwordPtr in
                salt.withUnsafeBytes { saltPtr in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        passwordPtr.bindMemory(to: Int8.self).baseAddress,
                        passwordData.count,
                        saltPtr.bindMemory(to: UInt8.self).baseAddress,
                        salt.count,
                        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                        pbkdf2Iterations,
                        derivedPtr.bindMemory(to: UInt8.self).baseAddress,
                        32
                    )
                }
            }
        }
        precondition(status == kCCSuccess, "PBKDF2 failed: \(status)")
        return derived
    }

    static func hmacSHA256(key: Data, message: Data) -> Data {
        let symmetric = SymmetricKey(data: key)
        let mac = HMAC<SHA256>.authenticationCode(for: message, using: symmetric)
        return Data(mac)
    }

    static func randomBytes(_ count: Int) -> Data {
        var data = Data(count: count)
        let result = data.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, count, $0.baseAddress!) }
        precondition(result == errSecSuccess)
        return data
    }
}

File: TestDPLS-Windows/src/TestDPLS.Core/Crypto/DplsCrypto.cs
using System.Security.Cryptography;
using System.Text;

namespace TestDPLS.Crypto;

public static class DplsCrypto
{
    public const int Pbkdf2Iterations = 10_000;

    public static byte[] DeriveVerifier(string password, ReadOnlySpan<byte> salt)
    {
        var passwordBytes = Encoding.UTF8.GetBytes(password);
        return Rfc2898DeriveBytes.Pbkdf2(
            passwordBytes,
            salt.ToArray(),
            Pbkdf2Iterations,
            HashAlgorithmName.SHA256,
            32);
    }

    public static byte[] HmacSha256(ReadOnlySpan<byte> key, ReadOnlySpan<byte> message)
    {
        using var hmac = new HMACSHA256(key.ToArray());
        return hmac.ComputeHash(message.ToArray());
    }

    public static byte[] RandomBytes(int count)
    {
        var data = new byte[count];
        RandomNumberGenerator.Fill(data);
        return data;
    }
}
